flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,539 @@
+
+The data associated with CapCons would be stored in arrays, 
+so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path. 
+CapCons are a non-storable object (similar to AuthAccounts).
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+    /// The block height when the capability was created.
+    let issueHeight: UInt64
+   
+    /// The Type of the capability, i.e.: the T in Capability<T>.
+    let borrowType: Type
+   
+    /// The id of the related capability.
+    /// This is the UUID of the created capability.
+    /// All copies of the same capability will have the same UUID
+    let capabilityID: UInt64
+   
+    /// Is the capability revoked.
+    let isRevoked: Bool
+
+    /// Returns the targeted storage path of the capability.
+    fun target(): StoragePath
+   
+    /// Revoke the capability making it no longer usable.
+    /// When borrowing from a revoked capability the borrow returns nil.
+    fun revoke()
+   
+    /// Retarget the capability.
+    /// This moves the CapCon from one CapCon array to another.
+    fun retarget(target: StoragePath)
+
+}
+```
+
+Capabilities would expose the new field `let id: UInt64`, in addition to the account field they already expose.
+
+The capability related methods in the `AuthAccount` and in the `PublicAccount` would be moved to a `capabilities` namespace, 
+similar to how the contract methods are in the [contracts namespace](https://developers.flow.com/cadence/language/accounts).
+
+The `AuthAccount` would get new methods to create or get or iterate through capabilities,
+and to get or iterate through CapCons in order to manage capabilities. 
+Methods used for (un)linking would be removed as they are no longer needed.
+
+```cadence
+struct AuthAccount {
+    // ...
+    // removed: fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+    // removed: fun getLinkTarget(_ path: CapabilityPath): Path?
+    // removed: fun unlink(_ path: CapabilityPath)
+    // moved & renamed: fun getCapability<T>(_ path: CapabilityPath): Capability<T>
+    // moved & renamed: fun forEachPublic(_ function: ((PublicPath, Type): Bool))
+
+    let capabilities: AuthAccount.Capabilities
+
+    struct Capabilities {
+        /// get returns the capability at the public path, if one was stored there.
+        fun get<T>(_ path: PublicPath): Capability<T>?
+
+        /// borrow gets the capability at the given path, and borrows the capability if it exists.
+        /// Returns `nil` if the capability does not exist or cannot be borrowed using the given type.
+        /// The function is equivalent to `get(path)?.borrow()`.
+        fun borrow<T>(_ path: PublicPath): T?
+ 
+        /// For each iterates through all the public capabilities of the public account.
+        /// Returning false from the function stops the iteration.
+        fun forEach(_ function: ((PublicPath, Type): Bool))
```

Comment:
Is there any reason the parameters for the functions changed from `CapabilityPath` (which can be either a public or private path) to just `PublicPath` (only public paths)?

Should the parameters be changed back to `CapabilityPath`?

```suggestion
        /// get returns the capability at the public path, if one was stored there.
        fun get<T>(_ path: CapabilityPath): Capability<T>?

        /// borrow gets the capability at the given path, and borrows the capability if it exists.
        /// Returns `nil` if the capability does not exist or cannot be borrowed using the given type.
        /// The function is equivalent to `get(path)?.borrow()`.
        fun borrow<T>(_ path: CapabilityPath): T?
 
        /// For each iterates through all the public capabilities of the public account.
        /// Returning false from the function stops the iteration.
        fun forEach(_ function: ((CapabilityPath, Type): Bool))
```

Reply:
The reason was that the FLIP also suggests removing the private storage path altogether.

Reply:
Ah, I see, that makes sense. Regarding migrating existing uses of `CapabilityPath`: We can probably alias it to `PublicPath`?

Reply:
Added this to the list of discussion points